File: exercises/exercise-1-testing-java/src/main/java/io/miragon/example/application/service/bike/SendPaymentReminderService.java
package io.miragon.example.application.service.bike;

import io.miragon.example.application.port.inbound.bike.SendPaymentReminderUseCase;
import io.miragon.example.application.port.outbound.bike.BikeSubscriptionRepository;
import io.miragon.example.domain.bike.BikeSubscriptionId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class SendPaymentReminderService implements SendPaymentReminderUseCase {

    private static final Logger log = LoggerFactory.getLogger(SendPaymentReminderService.class);

    private final BikeSubscriptionRepository repository;

    public SendPaymentReminderService(BikeSubscriptionRepository repository) {
        this.repository = repository;
    }

    @Override
    public void sendPaymentReminder(BikeSubscriptionId subscriptionId) {
        var subscription = repository.find(subscriptionId);
        log.info("Sending payment reminder to {} for bike subscription {}",
            subscription.customerEmail().value(), subscription.id().value());
    }

}

File: exercises/exercise-1-testing-kotlin/src/main/kotlin/io/miragon/example/application/service/bike/SendPaymentReminderService.kt
package io.miragon.example.application.service.bike

import io.miragon.example.application.port.inbound.bike.SendPaymentReminderUseCase
import io.miragon.example.application.port.outbound.bike.BikeSubscriptionRepository
import io.miragon.example.domain.bike.BikeSubscriptionId
import mu.KotlinLogging
import org.springframework.stereotype.Service

@Service
class SendPaymentReminderService(
    private val repository: BikeSubscriptionRepository
) : SendPaymentReminderUseCase {

    private val log = KotlinLogging.logger {}

    override fun sendPaymentReminder(subscriptionId: BikeSubscriptionId) {
        val subscription = repository.find(subscriptionId)
        log.info { "Sending payment reminder to ${subscription.customerEmail.value} for bike subscription ${subscription.id.value}" }
    }
}
